rp/src/cli.rs

```diff
@@ -0,0 +1,462 @@
+use std::path::PathBuf;
+use std::{iter::Peekable, net::SocketAddr};
+
+use crate::exchange::{ExchangeOptions, ExchangePeer};
+
+pub enum Command {
+    GenKey {
+        private_keys_dir: PathBuf,
+    },
+    PubKey {
+        private_keys_dir: PathBuf,
+        public_keys_dir: PathBuf,
+    },
+    Exchange(ExchangeOptions),
+    Help,
+}
+
+enum CommandType {
+    GenKey,
+    PubKey,
+    Exchange,
+}
+
+#[derive(Default)]
+pub struct Cli {
+    pub verbose: bool,
+    pub command: Option<Command>,
+}
+
+fn fatal<T>(note: &str, command: Option<CommandType>) -> Result<T, String> {
+    match command {
+        Some(command) => match command {
+            CommandType::GenKey => Err(format!("{}\nUsage: rp genkey PRIVATE_KEYS_DIR", note)),
+            CommandType::PubKey => Err(format!("{}\nUsage: rp pubkey PRIVATE_KEYS_DIR PUBLIC_KEYS_DIR", note)),
+            CommandType::Exchange => Err(format!("{}\nUsage: rp exchange PRIVATE_KEYS_DIR [dev <device>] [listen <ip>:<port>] [peer PUBLIC_KEYS_DIR [endpoint <ip>:<port>] [persistent-keepalive <interval>] [allowed-ips <ip1>/<cidr1>[,<ip2>/<cidr2>]...]]...", note)),
+        },
+        None => Err(format!("{}\nUsage: rp [verbose] genkey|pubkey|exchange [ARGS]...", note)),
+    }
+}
+
+impl ExchangePeer {
+    pub fn parse(args: &mut &mut Peekable<impl Iterator<Item = String>>) -> Result<Self, String> {
+        let mut peer = ExchangePeer::default();
+
+        if let Some(public_keys_dir) = args.next() {
+            peer.public_keys_dir = PathBuf::from(public_keys_dir);
+        } else {
+            return fatal(
+                "Required positional argument: PUBLIC_KEYS_DIR",
+                Some(CommandType::Exchange),
+            );
+        }
+
+        while let Some(x) = args.peek() {
+            let x = x.as_str();
+
+            // break if next peer is being defined
+            if x == "peer" {
+                break;
+            }
+
+            let x = args.next().unwrap();
+            let x = x.as_str();
+
+            match x {
+                "endpoint" => {
+                    if let Some(addr) = args.next() {
+                        if let Ok(addr) = addr.parse::<SocketAddr>() {
+                            peer.endpoint = Some(addr);
+                        } else {
+                            return fatal(
+                                "invalid parameter for listen option",
+                                Some(CommandType::Exchange),
+                            );
+                        }
+                    } else {
+                        return fatal(
+                            "listen option requires parameter",
```

Comment:
change "listen" to endpoint

Reply:
Resolved in fc1d4cc06023ceb0a9cf6362aed5930f1d2b2af4